javalib/src/main/scala/java/util/concurrent/ConcurrentSkipListSet.scala

```diff
@@ -0,0 +1,404 @@
+    extends AbstractSet[E]
+    with NavigableSet[E]
+    with Cloneable
+    with Serializable { self =>
+
+  def this() =
+    this(ConcurrentSkipListSet.defaultOrdering[E], false)
+
+  def this(comparator: Comparator[E])  = 
+    this(Ordering.comparatorToOrdering(comparator), true)
+
+  def this(sortedSet: SortedSet[E]) = {
+  	this(Ordering.comparatorToOrdering(sortedSet.comparator()), true)
+    addAll(sortedSet)
+  }
+
+  def this(collection: java.util.Collection[E]) = {
+  	this(ConcurrentSkipListSet.defaultOrdering[E], false)
+    addAll(collection)
+  }
+
+  def this(comparator: Comparator[E], collection: Collection[E]) = {
+    this(Ordering.comparatorToOrdering(comparator), true)
+    addAll(collection)
+  }
+
+  implicit object BoxOrdering extends Ordering[Box[E]] {
+    def compare(a: Box[E], b:Box[E]): Int = ordering.compare(a.inner, b.inner)
+  }
+
+  private val inner: mutable.SortedSet[Box[E]] = new mutable.TreeSet[Box[E]]()
+
+  override def isEmpty(): Boolean =   
+    inner.isEmpty
+
+  def iterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator
+
+      var actual: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        actual = Some(iter.next().inner)
+        actual.get
+      }
+
+      def remove(): Unit = actual.map(self.remove(_))
+    }
+  }
+
+  override def add(e: E): Boolean =
+    if (e === null) throw new NullPointerException()
+    else inner.add(Box(e))
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator()
+    var changed = false
+    while (iter.hasNext)
+      changed = add(iter.next()) || changed
+    changed
+  }
+
+  override def clear(): Unit =
+    inner.clear() 
+
+  override def clone(): ConcurrentSkipListSet[E] =
+    new ConcurrentSkipListSet(this)
+
+  def comparator(): Comparator[E] = {
+    if (!provided) null
+    else  {
+      new Comparator[E] {
+        def compare(a: E, b: E) = ordering.compare(a, b)
+      }
+    }
+  }
+ 
+  override def contains(o: Any): Boolean =
+    inner.contains(Box(o.asInstanceOf[E]))
+
+  def descendingIterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator.toIndexedSeq.reverse.iterator
+
+      var actual: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        actual = Some(iter.next().inner)
+        actual.get
+      }
+
+      def remove(): Unit = actual.map(self.remove(_))
+    }
+  }
+
+  def descendingSet(): NavigableSet[E] = {
+    val descSetFun = () => {
+      val inverseOrdering = 
+        new Ordering[Box[E]] {
+         def compare(a: Box[E], b: Box[E]) = self.ordering.compare(b.inner,a.inner)
+        }
+      val retSet = new mutable.TreeSet[Box[E]]()(inverseOrdering)
+      retSet.addAll(inner)
+      retSet
+    }
+    new ConcurrentSkipListSet.NavigableView(this, descSetFun)
```

Comment:
Unfortunately there is the need of recomputing(recursively) the actual
slice of list for any read operation, even knowing bounds and so on.
Probably the only way to avoid recreation is to maintain a "status
variable"  that will track if original is updated or not and than recompute
or not the actual slice.
On Jun 2, 2015 15:11, "Sébastien Doeraene" <notifications@github.com> wrote:

> In javalib/src/main/scala/java/util/concurrent/ConcurrentSkipListSet.scala
> <https://github.com/scala-js/scala-js/pull/1681#discussion_r31520038>:
>
> > +
> > +      def remove(): Unit = actual.map(self.remove(_))
> > +    }
> > +  }
> > +
> > +  def descendingSet(): NavigableSet[E] = {
> > +    val descSetFun = () => {
> > +      val inverseOrdering =
> > +        new Ordering[Box[E]] {
> > +         def compare(a: Box[E], b: Box[E]) = self.ordering.compare(b.inner,a.inner)
> > +        }
> > +      val retSet = new mutable.TreeSet[Box[E]]()(inverseOrdering)
> > +      retSet.addAll(inner)
> > +      retSet
> > +    }
> > +    new ConcurrentSkipListSet.NavigableView(this, descSetFun)
>
>  Modifications on NavigableSet are all backed recursively to original one.
>
> Right, so why not do the same for accesses? And get rid of the huge
> inefficiency of copying everything every time.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scala-js/scala-js/pull/1681/files#r31520038>.
>